llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -989,15 +989,47 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
 /// Try to simplify recipe \p R.
 static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
   using namespace llvm::VPlanPatternMatch;
-  // Try to remove redundant blend recipes.
+
   if (auto *Blend = dyn_cast<VPBlendRecipe>(&R)) {
-    VPValue *Inc0 = Blend->getIncomingValue(0);
+    // Try to remove redundant blend recipes.
+    SmallSet<VPValue *, 4> UniqueValues;
+    if (Blend->isNormalized() || !match(Blend->getMask(0), m_False()))
+      UniqueValues.insert(Blend->getIncomingValue(0));
     for (unsigned I = 1; I != Blend->getNumIncomingValues(); ++I)
-      if (Inc0 != Blend->getIncomingValue(I) &&
-          !match(Blend->getMask(I), m_False()))
-        return;
-    Blend->replaceAllUsesWith(Inc0);
+      if (!match(Blend->getMask(I), m_False()))
+        UniqueValues.insert(Blend->getIncomingValue(I));
+
+    if (UniqueValues.size() == 1) {
+      Blend->replaceAllUsesWith(*UniqueValues.begin());
+      Blend->eraseFromParent();
+      return;
+    }
+
+    if (Blend->isNormalized())
+      return;
+
+    // Normalize the blend so its first incomming value is used as the initial
+    // value with the others blended into it.
+
+    unsigned StartIndex = 0;
```

Comment:
Given that some of the code above already assumes the start index is 0 and hard-codes it (i.e. `!match(Blend->getMask(0), m_False())`) is it necessary to have a constant here? Alternatively, you could be consistent and declare this as a `const unsigned StartIndex = 0;` higher up and use it everywhere?

Reply:
There's no relationship between `StartIndex` and the code above.  The code above is walking through all operands hence starting at 0, whereas the intent of StartIndex is to capture the new start value for the normalised blends.

There's an associated PR (https://github.com/llvm/llvm-project/pull/104019) that adds code to update `StartIndex` which is why it's declared where it is and not a const.  I can add the const if you prefer but it will be short lived.

Reply:
I think this is just laying the ground for the next patch, which tries to find a better StartIndex before normalizing based on that.

I did wonder if it was better to just leave all the logic until that patch, and just remove the mask operand in this patch.

Reply:
OK fair enough. I only looked at this patch without looking at the follow-on.